Camera services arm disarm including Netgear Arlo

viswa-swami opened the pull request:
## Description:
Camera component needs a service called "Arm/Disarm" to be able to arm a camera on demand or via automation.  Similarly for disarm.

Arm: It will enable motion detection and also record the video/snapshot depending on camera setting
Disarm: It will disable motion detection and will not record anything.

The arguments (entity_id) is optional so that with just 'arm' or 'disarm' call, user can arm/disarm all the cameras in a shot.  Otherwise provide the specific entity_id of camera to arm/disarm if camera supports it.

This goes hand-in-hand with arlo camera where home assistant talks to the Arlo base station (Note that, for Arlo, we need to talk to base station/hub to control the cameras and not the cameras itself) to arm/disable all the cameras at once, for example.

These services are applicable to Foscam cameras too. For example, in my home setup, I currently have a shell command to arm/disarm the camera whenever we are not in home.  This shell command can be avoided if the arm/disarm service is available in camera component itself.  I can add the arm/disarm functionality for foscam cameras also once this codeset is done/committed/pushed.

I am not sure if we need to update the documentation anywhere on this concern.  If there is any, let me know and i can update the same.

These status can be used to create a camera based sensor to indicate whether the cameras are armed or disarmed.  That is next once these code are done.

**Related issue (if applicable):** fixes #<home-assistant issue number goes here>

**Pull request in [home-assistant.github.io](https://github.com/home-assistant/home-assistant.github.io) with documentation (if applicable):** home-assistant/home-assistant.github.io#<home-assistant.github.io PR number goes here>

## Example entry for `configuration.yaml` (if applicable):
```yaml

```

## Checklist:

If user exposed functionality or configuration variables are added/changed:
  - [ ] Documentation added/updated in [home-assistant.github.io](https://github.com/home-assistant/home-assistant.github.io)

If the code communicates with devices, web services, or third-party tools:
  - [ ] Local tests with `tox` run successfully. **Your PR cannot be merged unless tests pass**
  - [ ] New dependencies have been added to the `REQUIREMENTS` variable ([example][ex-requir]).
  - [ ] New dependencies are only imported inside functions that use them ([example][ex-import]).
  - [ ] New dependencies have been added to `requirements_all.txt` by running `script/gen_requirements_all.py`.
  - [ ] New files were added to `.coveragerc`.

If the code does not interact with devices:
  - [ ] Local tests with `tox` run successfully. **Your PR cannot be merged unless tests pass**
  - [ ] Tests have been added to verify that the new code works.

[ex-requir]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard.py#L14
[ex-import]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard.py#L54

viswa-swami committed:
Added camera service calls to arm/disarm the cameras. Entity id is optional so that with a single call we can arm all the cameras or specify a particular entity id to arm if applicable and possible in that camera type.

viswa-swami committed:
Added camera service calls to arm/disarm the cameras. Entity id is optional so that with a single call we can arm all the cameras or specify a particular entity id to arm if applicable and possible in that camera type.

viswa-swami committed:
Added camera service calls to arm/disarm the cameras. Entity id is optional so that with a single call we can arm all the cameras or specify a particular entity id to arm if applicable and possible in that camera type.

viswa-swami committed:
Fixed the spaces and indentation related issues that houndci found

viswa-swami committed:
Fixed the spaces and indentation related issues that houndci found

viswa-swami committed:
Missed the const file which has the macros defined.

viswa-swami committed:
Fixed the CI build error

viswa-swami committed:
Fixed the CI build error because of unused variable in exception case

viswa-swami commented:
I just now read and found that I can edit the arlo component page and add the arm/disarm service.  Please suggest if that is required to be added there or anywhere else would be appropriate.

pvizeli commented:
I think it would be more sense to have also a `alarm_control_panel ` platform for arlo as all hack into camera.

pvizeli reviewed (changes requested):
Please use `alarm_control_panel` platform for arlo instead to copy that into camera platform.

viswa-swami commented:
Arm/Disarm is just the same wording used between alarm and camera, but why would we need the entire control panel for a simple arm/disarm? This would seem to keep the component clean instead of adding a full fledged panel right?

viswa-swami commented:
This arm/disarm service in camera component is quite useful in arming/disarming the cameras directly as a service or via automation then using it via alarm control panel platform.

viswa-swami commented:
@pvizeli this is a generic arm/disarm for all cameras including foscam, amcrest and Arlo too. Currently I am using them by adding a separate python script and calling them using the shell_command platform/service.

Adding this would make it self reliant instead of each user needing to add scripts on their own and calling them.

pvizeli commented:
I'll speak with @balloob. But you need write tests for that and add this to demo camera too. It also not allow to catch AttributeError on that point. So it need more work for add this into base component.

viswa-swami commented:
@pvizeli thank you.  I have some clarifications based on your comment.  Can you please review and suggest accordingly.

Regarding tests:
Where do i need to write the tests ? Is there any format or document or any link where i should update them ?

Demo camera:  
I am not very clear on what to add for the demo camera.  I am assuming it is the components/camera/demo.py.  Is that right? If so, I can add these `async_arm()` and `async_disarm()` to this as an empty function !?

AttributeError:
I added this exception because I hadn't added the `async_arm()` and `async_disarm()` for other cameras at that time.  I have foscam and arlo cameras at my place.  I can add the code for foscamm camera too.  But for other camera models, I can add an empty function `async_arm()` and `async_disarm()`.  That way, anyone who has those camera models can add them !?

balloob commented on homeassistant/components/camera/__init__.py:
```diff
@@ -125,6 +199,16 @@ def brand(self):
         return None
 
     @property
+    def status(self):
```
Why add something new if we already have `state` that does something similar?

balloob commented on homeassistant/components/camera/arlo.py:
```diff
@@ -90,3 +93,22 @@ def model(self):
+    @asyncio.coroutine
+    def async_arm(self):
+        """Camera arm."""
+        self._base_stn.mode = "armed"
```
You are not allowed to do I/O inside async methods.

viswa-swami replied:
Is there any example as to how to handle I/O, if we cannot do it inside the async methods ?

balloob commented on homeassistant/components/camera/arlo.py:
```diff
@@ -90,3 +93,22 @@ def model(self):
+    @asyncio.coroutine
+    def async_disarm(self):
+        """Camera disarm."""
+        self._base_stn.mode = "disarmed"
```
Same

viswa-swami replied:
Is there any example as to how to handle I/O, if we cannot do it inside the async methods ?

balloob replied:
```python
def set_mode(mode):
    self._base_stn.mode = mode

yield from hass.async_add_job(set_mode, "disarmed")
```

balloob commented on homeassistant/const.py:
```diff
@@ -315,6 +315,9 @@
 SERVICE_HOMEASSISTANT_STOP = 'stop'
 SERVICE_HOMEASSISTANT_RESTART = 'restart'
 
+SERVICE_ARM = 'arm'
```
Please keep these in the component

viswa-swami replied:
I didn't see a const.py inside the component/ directory.  Shall I just create a const.py inside components/ directory or shall I just add this inside the __init__.py or something !?

balloob replied:
With inside the component I mean define them inside `camera/__init__.py`

balloob commented on homeassistant/components/camera/__init__.py:
```diff
@@ -92,6 +125,47 @@ def update_tokens(time):
+                elif service.service == SERVICE_DISARM:
+                    yield from camera.async_disarm()
+            except AttributeError:
+                pass
```
We should not just let this pass. We should also not try to update a camera that didn't had an attribute.

viswa-swami replied:
I can check if the camera has that attribute using `hasattr()` and then call that if it has that attribute.

balloob reviewed (commented):
Adding generic services to a component means that you will have to update the demo camera platform and write tests to make sure that everything works.

I am not 100% convinced that arm/disarm functionality is something that most camera's have. For example with Nest Cam you mute notifications, alerts are always on. 

That's why it might be easier for now to add platform specific services: `arlo_arm`, `arlo_disarm`. You would register these services inside your platform, that way you don't have to write tests, update the demo platform etc.

viswa-swami commented:
Thanks @balloob for your suggestions.  

The reasoning behind why I did it as a generic component was because I have foscam cameras inside my home and Arlo for outdoor.  For privacy reasons, I keep the foscam cameras turned off when we are home, and turn on/enable motion detection only when we are not in home.

Currently I had to write a separate python script and call it via shell_command for automation.  To avoid it, I created a generic camera arm/disarm service so that I could just invoke that service instead of separate script.

I was not sure about nest cameras, but cameras like foscam, amcrest and Arlo of course has these capabilities of enabling and disabling motion/sound detection via APIs.

If you think it will be useful for others too as a service, please suggest and I can work on adding the demo and tests.  But if you think it would be an Overkill at the moment, I can add this just for Arlo.

@balloob, can you please recommend which route to go?

viswa-swami commented:
In parallel, I am also modifying the code based on your review comments.  Currently I have removed the AttributeError related code, as well as moved the I/O out of the async functions based on the "Working with Async" documentation in home-assistant.io.  I will update the PR once that is ready and tested in my home.

balloob commented:
For Foscam you're talking about both turning the camera on and turning on the motion detection. I would expect an 'arm' service to only turn on the motion detection. Turning a device on/off is a separate command. In that case I also wonder if `arm` and `disarm` are the wrong names. What about `set_motion_detection` ?

In that case we also would have `detecting_motion` be a new property that returns a boolean and is part of the device attributes.

Like Pascal said, if you are going to add this to the camera component, you will also have to update the demo camera platform. The methods will have to fake the data on the instance and report if motion reporting is on as if it is backed by a device. See the light demo for an example.

viswa-swami commented:
Sorry for the confusion.  I was explaining my use case as to how I operate the camera in my home.

This `arm` and `disarm` only takes care of enabling the motion detection for the given camera entity.  Based on your comment, instead of arm and disarm, shall I rename the service also to "enable_motion_detection" and "disable_motion_detection" respectively ?

The "status" that I have added does this purpose of indicating whether the camera is armed or disarmed.  It can potentially be used for a sensor also.

I have already moved the macros to camera/__init__.py as well as added to the demo.py.  I will update the PR a bit later.  I have also added code to remove the AttributeError and verified it here.

viswa-swami committed:
Updating the arlo code based on comment from @balloob. Changed the arm and disarm to enable_motion_detection and disable_motion_detection respectively. Similarly fixed the AttributeError handling. Added dummy code to the demo camera also. Moved out the definitions in const.py into the camera __init__ file

viswa-swami committed:
Fixed the comments posted by houndci-bot

viswa-swami committed:
Fixed the comments posted by houndci-bot

viswa-swami committed:
Fixed the comments posted by houndci-bot

viswa-swami commented:
This is how the arlo camera entity description shows as:

```
	entity_picture: /api/camera_proxy/camera.back_yard?token=69c1f02d830508f0760031bc756e1546a6b3653c1bc921068d571ca55b20d9f6
access_token: 69c1f02d830508f0760031bc756e1546a6b3653c1bc921068d571ca55b20d9f6
motion_detection: Disabled
model_name: VMC3030
friendly_name: Back Yard
brand: Netgear Arlo
```
The 'motion_detection' field is the one that has been added now.  It will show 'Enabled' when the respective service is invoked.

viswa-swami commented:
@balloob I have made the changes as per your comments and have posted them here.  I have changed the names from 'arm', 'disarm' to 'enable_motion_detection' and 'disable_motion_detection' respectively.  Similarly updated the demo code also.  Please review and let me know your suggestions. 

Once this is completed, I have the foscam camera too. I can add the motion enable/disable code for foscam also.

viswa-swami committed:
Fixed the comments posted by travis-ci integration bot

viswa-swami committed:
Fixed the comments posted by travis-ci integration bot

viswa-swami committed:
Fixed the comments posted by travis-ci integration bot for demo.py: expected 2 lines, found 1

viswa-swami committed:
Updated code in camera __init__.py to use the get function instead of directly calling the member in the structure.

viswa-swami committed:
Updated code in camera __init__.py

balloob commented on homeassistant/components/camera/__init__.py:
```diff
@@ -125,6 +203,11 @@ def brand(self):
         return None
 
     @property
+    def get_motion_detection_status(self):
```
Please rename to `motion_detection_enabled`

viswa-swami replied:
done

balloob commented on homeassistant/components/camera/arlo.py:
```diff
@@ -90,3 +94,30 @@ def model(self):
+    @property
+    def get_motion_detection_status(self):
+        """Camera Motion Detection Status."""
+        return self._motion_status
```
This should return a boolean.

viswa-swami replied:
ok, i will change this code to return a boolean and post for PR

balloob commented on homeassistant/components/camera/arlo.py:
```diff
@@ -90,3 +94,30 @@ def model(self):
+        return self._motion_status
+
+    @asyncio.coroutine
+    def async_enable_motion_detect(self):
```
`async_enable_motion_detection`

balloob commented on homeassistant/components/camera/arlo.py:
```diff
@@ -90,3 +94,30 @@ def model(self):
+        """Camera arm."""
+        self._motion_status = MOTION_ENABLED
+        self.hass.async_add_job(self.async_update_ha_state())
+        self.hass.async_add_job(self.async_update())
```
Don't call update here. Instead overwrite the property `should_poll` to return `True`. That way update will be done automatically after the service is called (as per this PR)

balloob commented on homeassistant/components/camera/arlo.py:
```diff
@@ -90,3 +94,30 @@ def model(self):
+    def async_disable_motion_detect(self):
+        """Camera disarm."""
+        self._motion_status = MOTION_DISABLED
+        self.hass.async_add_job(self.async_update_ha_state())
```
Same, remove this.

balloob commented on homeassistant/components/camera/arlo.py:
```diff
@@ -90,3 +94,30 @@ def model(self):
+        self.hass.async_add_job(self.async_update())
+
+    @asyncio.coroutine
+    def async_update(self):
```
Async update is reserved to update the state of the entity from the actual device. You should not use it to write to the base station.

viswa-swami replied:
how should i update the base station then ?  I am not very clear from the "Working with Async" documentation as to where to do the actual device communication !  Can you please clarify ?

balloob commented on homeassistant/components/camera/arlo.py:
```diff
@@ -90,3 +94,30 @@ def model(self):
+    def async_update(self):
+        """Perform the I/O operation with camera."""
+        if self._motion_status == MOTION_ENABLED:
+            self._base_stn.mode = ARLO_MODE_ARMED
```
You are not allowed to do I/O inside a coroutine.

viswa-swami replied:
how should i update the base station then ?  I am not very clear from the "Working with Async" documentation as to where to do the actual device communication !  Can you please clarify ?

viswa-swami replied:
should i be creating a function to write into the base station and then yield from that function from async_update() ?

viswa-swami replied:
should i move the code where i set into the base_stn to a separate function and then call that function using yield from the async_update().  I am going through other examples like media_players/kodi.py, and they seem to be doing something similar to this !

balloob commented on homeassistant/components/camera/demo.py:
```diff
@@ -5,25 +5,32 @@
         super().__init__()
+        self._parent = hass
         self._name = name
+        self._motion_status = MOTION_DISABLED
```
boolean.

viswa-swami replied:
changed to boolean

balloob commented on homeassistant/components/camera/demo.py:
```diff
@@ -38,3 +45,30 @@ def camera_image(self):
+    def async_update(self):
+        """Perform the I/O operation with camera."""
+        if self._motion_status == MOTION_ENABLED:
+            _LOGGER.info("Enable Motion detection for this camera here")
```
This is wrong. You have already updated for the demo by setting the instance variable. There is no need to add an `async_update`.

You do however also want to add a `should_poll` property and set it to True on this entity.

viswa-swami replied:
I have added the should_poll property override and removed the async_update code from the demo.py

balloob commented on homeassistant/components/services.yaml:
```diff
@@ -402,3 +402,18 @@ eight_sleep:
         description: Duration to heat at the target level in seconds.
         example: 3600
+
+camera:
```
Please remove these, as they are already present in `camera/services.yaml`.

viswa-swami replied:
done

balloob reviewed (changes requested):
You will need to add a test to make sure that the methods on the camera component can successfully change the motion detected status of the demo camera.

viswa-swami commented on homeassistant/components/camera/__init__.py:
```diff
@@ -12,23 +12,31 @@
 
+SERVICE_EN_MOTION = 'enable_motion_detection'
+SERVICE_DISEN_MOTION = 'disable_motion_detection'
+MOTION_ENABLED = 'Enabled'
```
ok, i will change this code to return a boolean and post for PR

viswa-swami commented:
@balloob Thanks for the comments/suggestions.  I have commented with some of my questions regarding handling the base station I/O activity.  Can you please clarify that so that i can update the code accordingly.

Regarding the tests, Where should i write the tests? Is there a page in homeassistant.io or something where I should write the tests ?  Is there any example or reference that i can refer to, so that i can use the same format ?

viswa-swami commented on homeassistant/components/camera/__init__.py:
```diff
@@ -43,6 +51,36 @@
+})
+
+
+def en_motion_detection(hass, entity_id=None):
```
done

viswa-swami commented on homeassistant/components/camera/__init__.py:
```diff
@@ -43,6 +51,36 @@
+
+
+@callback
+def async_en_motion_detection(hass, entity_id=None):
```
done

viswa-swami commented on homeassistant/components/camera/__init__.py:
```diff
@@ -43,6 +51,36 @@
+        DOMAIN, SERVICE_EN_MOTION, data))
+
+
+def disen_motion_detection(hass, entity_id=None):
```
done

viswa-swami commented on homeassistant/components/camera/__init__.py:
```diff
@@ -43,6 +51,36 @@
+
+
+@callback
+def async_disen_motion_detection(hass, entity_id=None):
```
done

viswa-swami commented:
@balloob I reviewed the `media_player/kodi.py` and saw that, from an async coroutine in the` init.py`, they call the device specific '`set_volume_level`' code.  The `kodi.py` implements this function and sets the volume in the device.

So, based on that, can i do this:
1. remove the `async_update()` code.
2. in the async_enable_motion_detection(), remove the` @asyncio.coroutine` and talk to the base_station there.
3. This code is actually called from a async coroutine in the `init.py` which yields to this function.

```
    def async_enable_motion_detection(self):
        """Enable the motion detection in the camera (Arm mode)."""
        self._motion_status = True
        self._base_stn.mode = ARLO_MODE_ARMED

    def async_disable_motion_detection(self):
        """Disable the motion detection in the camera (Disarm mode)."""
        self._motion_status = False
        self._base_stn.mode = ARLO_MODE_DISARMED
```

viswa-swami commented:
@balloob I apologize for asking too many questions... I am not very much clear on the async documentation, and hence clarifying my questions with you as and when i implement your recommendations,....

balloob commented:
try this:

```python
    def async_enable_motion_detection(self):
        """Turn the entity on.

        This method must be run in the event loop and returns a coroutine.
        """
        return self.hass.async_add_job(self.enable_motion_detection)

    def enable_motion_detection(self):
        """Turn the entity off."""
        self._motion_status = True
        self._base_stn.mode = ARLO_MODE_ARMED
```

You can then in the service handler still yield from the async method.

balloob commented:
`async_add_job` will return a coroutine, what is what we're yielding from.

The `enable_motion_detection` doesn't have annotations and thus will be run in a thread by `async_add_job`.

Do you still need the line `self._motion_status = True` ? Shouldn't that be fetched inside `update` ? That way if motion detection is enabled via the app, it also is reflected in Hass.

viswa-swami commented:
@balloob Thanks a lot for your reference code and explanation.  Now it makes sense and able to understand the flow.

Regarding the self._motion_status, Yes, I think we still need that at the least for now.  

pyarlo which is what is being used for the arlo component, hasn't implemented the Get method yet.  The setter function is setting the mode correctly, but the get function is returning None (as of latest pyarlo 0.0.4).  We can have this for now.

I will try to add that into pyarlo once this is done.  If that is completed, then we can probably use that later version as requirement and remove this _motion_status.  Then it will sync properly between app/website & hass.

Only drawback now is that, if the cameras are already disarmed (no motion detection), Hass may send the Disable motion detection again (as part of automation for example), but that will not affect the camera state based on my experiments in my setup here.

With this current method, at least, it will work properly with Hass and we can enable/disable the motion detection.

viswa-swami commented:
@balloob Can you also please suggest where I should add the tests for this PR that you requested earlier ?  Thanks in Advance for all your help and suggestions.

viswa-swami commented:
@balloob Regarding the tests, I understand that I should add it under home-assistant/tests/components/camera/ directory, but is there any already available file that i should add it to?

viswa-swami committed:
Posting the updated code for PR based on @balloob's suggestions/recommendations

viswa-swami committed:
Removed the arlo reference from demo code. Copy-paste error

viswa-swami committed:
Removed the unused import found by hound bot

viswa-swami committed:
Expected 2 lines before function, but found only 1.

balloob commented:
Oh interesting, there is currently no test file yet for the demo camera. You will have to add a new file `test_demo.py` in the `tests/components/camera` folder.

balloob commented on homeassistant/components/arlo.py:
```diff
@@ -22,6 +22,8 @@
 DOMAIN = 'arlo'
 
 DEFAULT_BRAND = 'Netgear Arlo'
+ARLO_MODE_ARMED = 'armed'
```
Please move these constants to the `camera/arlo.py` file.

balloob commented:
You can have a look at the `test_local_file.py` or `test_generic.py` tests as inspiration. 

Example test (untested):

```python
from asyncio import coroutine
from homeassistant.components import camera
from homeassistant.setup import async_setup_component

@asyncio.coroutine
def test_motion_detection(hass):
    """Test motion detection services."""
    # Setup platform
    yield from async_setup_component(hass, 'camera', {
        'camera': { 'platform': 'demo' }
    })

    # Fetch state and check motion detection attribute
    state = hass.states.get('camera.demo_camera')
    assert not state.attributes.get('motion_detection')

    # Call service to turn on motion detection
    camera.enable_motion_detection(hass, 'camera.demo_camera')
    yield from hass.async_block_till_done()

    # Check if state has been updated.
    state = hass.states.get('camera.demo_camera')
    assert state.attributes.get('motion_detection')
```

viswa-swami committed:
Based on @balloob's comments, moved these constants to the camera/arlo.py

viswa-swami commented:
@balloob Thanks for your suggestions.

I have committed my changes with your suggestion to move the constants out into camera/arlo.py.

Interesting... I will look at this test file that you mentioned and try to add the test_demo.py for camera.  How do I run the test to verify if it is working before I push it for PR later when done ?

balloob commented:
https://home-assistant.io/developers/development_testing/

viswa-swami commented:
👍

viswa-swami committed:
Added test_demo.py to test the motion enabled and motion disabled in camera component

viswa-swami committed:
Fixing issues found by houndci-bot

viswa-swami committed:
Fixing issues found by houndci-bot

viswa-swami commented:
@balloob I have added a test_demo.py based on your example (thanks for your example reference code).  I ran the 'tox' command and it succeeded for this file.

Please review and suggest accordingly.

balloob commented on homeassistant/components/camera/arlo.py:
```diff
@@ -90,3 +94,35 @@ def model(self):
+        """Set the mode in the base station."""
+        self._base_stn.mode = mode
+
+    def async_enable_motion_detection(self):
```
These methods need to be moved to the base entity with the sync version raising `NotImplementedError`. That's how we do it in other places in HASS too. See for example climate: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/climate/__init__.py#L608-L617

balloob commented on homeassistant/components/camera/demo.py:
```diff
@@ -38,3 +42,29 @@ def camera_image(self):
+        """Camera Motion Detection Status."""
+        return self._motion_status
+
+    def async_enable_motion_detection(self):
```
Same, should be removed in favor of a version on the base entity.

viswa-swami committed:
Fixing the code as per @balloob's suggestions

viswa-swami committed:
Fixing the code as per @balloob's suggestions

viswa-swami commented:
@balloob Posted the PR after verifying the functionality with actual code and that was working fine !!! Couldn't seem to figure out why the test_demo is failing though !  Looking at the demo code to see if something got missed !

viswa-swami committed:
Fixing the test_demo failure. Tried to rewrite a base function to enable the motion in __init__.py and missed to add it to as a job.

viswa-swami committed:
Fixing the hound bot comment

viswa-swami commented:
@balloob Figured out the problem, and fixed it.  Have submitted the latest code for PR.  

Please review and let me know your suggestions if any.

Thanks for your help and patience... :)

tchellomello commented:
@viswa-swami FYI https://pypi.python.org/pypi/pyarlo/0.0.5 is available :+1:

balloob committed:
Merge branch 'dev' into arlo_services_arm_disarm

balloob committed:
Update arlo.py

balloob committed:
Update arlo.py

balloob commented:
Cherry-picked for 0.48

viswa-swami commented:
Thank you @balloob for including this in the upcoming release.
